ProjFS.Mac/PrjFSKext/VnodeUtilities.cpp

```diff
@@ -10,8 +10,8 @@ FsidInode Vnode_GetFsidAndInode(vnode_t vnode, vfs_context_t context)
 {
     vnode_attr attrs;
     VATTR_INIT(&attrs);
-    // TODO: check this is correct for hardlinked files
-    VATTR_WANTED(&attrs, va_fileid);
+    // va_linkid is unique per hard link, va_fileid only identifies the content
+    VATTR_WANTED(&attrs, va_linkid);
```

Comment:
For my own understanding, is it possible for the linkid of a vnode to change without the vnode's vid changing?

While working on the vnode cache changes I had to do the following:

```
    // Currently delete actions invalidate the cache entry to handle the hardlink+delete rename scenario.  If we do not invalidate
    // the entry we'll find the root for the newly created hardlink and we need to find the root of the path of the file being deleted.
    // Testing has shown that looking up the root again has consistently yielded the root of the file being deleted.
```

Do you know if this could have been avoided if I also stored/compared linkid when looking up an entry in the cache?

Reply:
I realised my experience with the edge cases here were somewhat out of date, so I've just run through a few scenarios on both APFS and HFS+ on macOS 10.13.6.

On HFS+:
 * When you create a file, its linkid is equal to its fileid.
 * As soon as you create one or more additional links to a file, each hard link, including the initial one, gets its own separate linkid. So even the original link's ID is changed. The fileid is the same for all of them though.
 * Which link (and indeed linkid) a particular vnode is associated with is not fixed; there seems to be only one vnode for a multi-hardlinked file at a time, and the linkid will fluctuate over time. (This is what I apparently misremembered, or possibly what has changed since 10.8~10.9.)

On APFS:

 * The linkid is apparently always the same as the fileid, even with multiple hard links. The path (and parent vnode) of the vnode keeps changing.

I believe `fsgetpath()` will always get the "correct" path to the file if given the appropriate linkid on HFS+, but on APFS I guess there's no way to distinguish between them.

Overall, this isn't great news for us on either file system; I'm fairly sure there's no way to guarantee we'll always end up with the "correct" path we're after even on HFS+, let alone APFS. This is pretty much orthogonal to the changes in this PR, incidentally. But it does mean that correctly handling hard links is going to be a tricky problem.

Reply:
> This is pretty much orthogonal to the changes in this PR, incidentally. But it does mean that correctly handling hard links is going to be a tricky problem.

Yep I agree thanks!  Seeing this change in your code got me wondering if there was something different I should be doing with hardlinks and the vnode cache.

Reply:
> Seeing this change in your code got me wondering if there was something different I should be doing with hardlinks and the vnode cache.

Nothing fully-baked, but some thoughts that might give you some further ideas:
 * Hard-linked files are easy to detect using the `va_nlink` attribute. If that value is 1, you have nothing to worry about. If it's not 1, `vnode_getparent()` will give you *a* parent, possibly not the one you want. `fsgetpath()` might give you the path inside the provider's root, or it might not.
 * 3 central problems I can see with a multi-hardlinked file:
  1. *Identifying the root(s) where it's a member.* Even with the vnode cache, this is tricky as we don't have a way of iterating over all the links to the file, and vnodes get recycled. Perhaps we want to store a list of UUIDs of virtualisation roots in an xattr? With the vnode cache, we can at least cache this data so we don't take the hit for reading xattrs all the time.
  2. *How to handle various file operations in the case where it's a member of more than 1 root.* Is there any conceivable way such a file could sensibly be anything other than a "full" file? If not, we could *always* convert to full when a hardlink to outside the repo is made, and prevent hardlinking non-full files in offline roots. Unloading the prjfs kext still lets you mess with this.
  3. *In the provider, what to do when `fsgetpath()` (or a fileop handler arg) returns a path outside the root.* The best approach I can currently think of here is to track the inode numbers of all "full" files and look up inodes for multi-linked files there before using `fsgetpath()`.
  4.

Reply:
@pmj thanks for the suggestions!  I've added a link to your comment to related issue #337

Reply:
The latest iteration of commit 3f19b35 in this series has been adjusted with what I learned from that linkid vs fileid investigation.